### PR TITLE
Added secure string support for credentials

### PR DESCRIPTION
**Pull request by ckuhn203**
Issue  #1048

I added some extra comments and a helper function in the constants class as well. Obviously I removed my credentials before committing and turned the tests back off, but they passed. You can use this snippet to re-instate them for this change.

            return new SecureUsernamePasswordCredentials() 
            { 
                Username = StringToSecureString("username"), 
                Password = StringToSecureString("swordfish") 
            };

### CODE REVIEW

**Comment by nulltoken** [comment]
@ckuhn203 Thanks! Looks clean to me.

@ethomson @jamill @Therzok @bording Thoughts?

**Comment by ckuhn203** [comment]
One thing I will say before anyone merges this is that I did consider making the existing credentials and this one share a generic interface, but this was the simplest thing that worked. I figured "Copy paste once: ok. Copy paste twice: stop. Refactor."

**Comment by ckuhn203** [comment]
I might have been quick on the draw there. I'm thinking maybe `Username` should be a plain string.  That's the approach [Microsoft took with `NetworkCredential`](https://msdn.microsoft.com/en-us/library/dd783746%28v=vs.110%29.aspx). Thoughts?

**Comment by nulltoken** [comment]
>Thoughts?

Makes sense, indeed. Could you please amend your commit with this change?

File: LibGit2Sharp.Tests/TestHelpers/Constants.cs [hunk 1]
```diff
@@ -68,5 +71,21 @@ public static string BuildPath()
...
             return testWorkingDirectory;
         }
+
+        // To help with creating secure strings to test with.
+        private static System.Security.SecureString StringToSecureString(string str)
+        {
+            var chars = str.ToCharArray();
+
+            var secure = new System.Security.SecureString();
+            for (var i = 0; i < secure.Length; i++)
```
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/Constants.cs]: Shouldn't this be `chars.Length`?

**Comment by Therzok** [comment]
I'd say keep the Username as a secure string, tbh. The username can be sensitive enough data to brute force the password on some providers. :P

**Comment by Therzok** [comment]
Actually, nevermind, just saw that NetworkCredential _does_ use a string.

**Comment by ckuhn203** [comment]
> Shouldn't this be chars.Length?

Yes. It should have been. I would have thought something would have failed if the wrong password got sent in... Either way, I corrected that.

**Comment by ckuhn203** [comment]
Okay... I don't get it. I've run the tests locally several times. Results seem intermittent. Sometimes everything passes. Sometimes I get failures. Failures are never the same tests and if I simply re-run the failed tests, they pass. Is this a known issue?

**Comment by nulltoken** [comment]
I've restarted the build. And it passed. Giving a look at the previous AppVeyor logs, looked like a GitHub transient issue.

**Comment by nulltoken** [comment]
:cool:

Before we merge this in, could you please squash those commits together into a nice cohesive one? We tend not to merge review commits.

**Comment by ckuhn203** [comment]
Thanks @nulltoken. Much appreciated. 

Keep up the good work @all. Love the library. Thanks for all your hard work on it (and for helping me to contribute in my own small way)!

**Comment by nulltoken** [comment]
@ckuhn203 Every contribution is welcome. Yours was very thoughtful! Thanks a lot for it. :sparkles: :sparkles:

**Comment by nulltoken** [comment]
Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150516171636`

**Comment by ckuhn203** [comment]
:+1: thanks again. I'll keep my eye on the easy fixes.

**Pull request merged**